File: solutions/017.letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.js
const map = new Map([
  ['2', 'abc'],
  ['3', 'def'],
  ['4', 'ghi'],
  ['5', 'jkl'],
  ['6', 'mno'],
  ['7', 'pqrs'],
  ['8', 'tuv'],
  ['9', 'wxyz'],
])

const letterCombinations = digits => {
  if (digits === '') return []

  let res = []

  const bfs = (temp, left) => {
    if (left.length === 0) {
      res.push(temp)
      return
    }

    let options = map.get(left[0])

    for (let i = 0; i < options.length; i++) {
      let item = options[i]
      bfs(temp + item, left.slice(1, left.length))
    }
  }

  bfs('', [...digits])

  return res
}

File: solutions/017.letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        map = {
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqrs',
            '8': 'tuv',
            '9': 'wxyz',
        }
        if not digits:
            return []

        ret = []

        def traverse(tmp, pos):
            nonlocal digits

            if pos > len(digits):
                return

            if pos == len(digits):
                ret.append(tmp)
                return

            cur = map[digits[pos]]

            for c in cur:
                traverse(tmp + c, pos + 1)

        traverse('', 0)

        return ret
